### PR TITLE
Adaptive publishing improvements

### PR DESCRIPTION
**Pull request by boks1971**
- Do not disable layer 0
- Run a check for max subscribed quality 30 seconds after creating a new media track
- Ensure connection quality does not get dinged because of missing layers that are above max subscribed layer
- When a layer is not subscribed and it stops, it is possible that the layer is stopped intentionally by the client. In that case, reset the stream tracker so that a layer re-starts on first packet. Use a timer for that as there could be a delay between when the max subscribed quality is calculated and when the layer actually stops.

### CODE REVIEW

File: pkg/sfu/streamtracker.go [hunk 1]
```diff
@@ -121,6 +121,21 @@ func (s *StreamTracker) Stop() {
...
 }
 
+func (s *StreamTracker) Reset() {
+	s.Stop()
+
+	s.countSinceLast = 0
+	s.cycleCount = 0
+
+	s.initMu.Lock()
+	s.initialized = false
```
Comment (davidzhao) [inline pkg/sfu/streamtracker.go:131]: re-using the init logic during reset will trigger multiple `detectWorker` to start. seems that we don't want that.

File: pkg/sfu/receiver.go [hunk 2]
```diff
@@ -209,6 +216,21 @@ func (w *WebRTCReceiver) AddUpTrack(track *webrtc.TrackRemote, buff *buffer.Buff
...
+				// If the layer is expected to stop, reset tracker so that it starts up
+				// on the first packet of the layer on restart.
+				//
+				// As there may be packets from client before a layer expected to stop actually stops,
+				// use a window in which a status stopped is treated as an expected stop.
+				// If the stop happens outside the window, treat it like an unexpected stop.
+				//
+				maxExpectedLayer, ok := w.maxExpectedLayer.Load().(MaxExpectedSpatialLayer)
+				if ok {
+					if time.Since(maxExpectedLayer.setAt) < 30*time.Second && layer > maxExpectedLayer.layer {
```
Comment (davidzhao) [inline pkg/sfu/receiver.go:230]: instead of the time window, would it be better if we called a function on receiver right before resuming a layer on the server side? it seems more predictable and we could get rid of the timing here.

**Review by davidzhao — COMMENTED** [review]
looks good. a couple of comments

File: pkg/rtc/mediatrack.go [hunk 3]
```diff
@@ -119,6 +119,10 @@ func NewMediaTrack(track *webrtc.TrackRemote, params MediaTrackParams) *MediaTra
 	t.AddOnClose(t.closeChan)
 	go t.updateStats()
 
+	time.AfterFunc(30*time.Second, func() {
```
[the remark this answers is not in the document]
Comment (boks1971) [reply pkg/rtc/mediatrack.go:122]: Yeah, will move it. Making some changes to start this when a receiver is created. Also, adding a `Stop` on the timer just in case the track closes before 30 seconds.

File: pkg/sfu/streamtracker.go (hunk 1 above)
Comment (boks1971) [reply pkg/sfu/streamtracker.go:131]: `Reset` calls `Stop` first which closes `s.running` channel which will end the current `detectWorker`. Can you please explain how multiple `detectWorker`s will start?

File: pkg/sfu/receiver.go (hunk 2 above)
Comment (boks1971) [reply pkg/sfu/receiver.go:230]: Will look at moving this on expected start.

File: pkg/sfu/streamtracker.go (hunk 1 above)
Comment (davidzhao) [reply pkg/sfu/streamtracker.go:131]: maybe I'm overthinking this.. Stop() doesn't immediately stop the detector, which checks `IsRunning` periodically in a loop.. As long as we have a reasonable amount of time between Reset and receiving packets again (like how it's currently being used), it wouldn't be an issue.

when we move Reset to immediately before resume, then it feels like the timing issue is more prominent? perhaps it's still much longer than the detect loop.

File: pkg/sfu/receiver.go (hunk 2 above)
Comment (boks1971) [reply pkg/sfu/receiver.go:230]: Would this be some thing like the following?
- When max subscribed layer is decreasing, we do not have to do anything. If the client implements stopping layers, the layer will stop eventually and stream tracker will catch it and the rest flows like now.
- When max subscribed layer is increasing, two things need to be handled
   - If clients did not stop any layer, we do not need to reset the tracker.
   - If the clients did stop a layer, we need to reset.
   So, maybe, we cross check with currently available layers and reset layers which are not available now, but is expected to start shortly? That should cover both cases.
   
Does that make sense?

**Review by davidzhao — APPROVED** [review]

File: pkg/sfu/streamtracker.go (hunk 1 above)
Comment (boks1971) [reply pkg/sfu/streamtracker.go:131]: Got it. No overthinking. Thank you for catching these edge cases. These are the ones which will come back to bite at the most inopportune time :-). I will think about how to address this.

File: pkg/sfu/receiver.go (hunk 2 above)
Comment (davidzhao) [reply pkg/sfu/receiver.go:230]: ah yes, that's a good case you are mentioning. we'd need to check the current layer state to see if a reset is necessary. 👍

Comment (boks1971) [reply pkg/sfu/receiver.go]: @davidzhao Have addressed your comments in this commit - https://github.com/livekit/livekit-server/pull/295/commits/973e2b400381018afe64501b34f3334da67093c1. Also set the first update to be after 10 seconds. Seemed like a good enough wait to do that.

Comment (boks1971) [reply pkg/sfu/receiver.go]: Added this commit as well for proper locking - https://github.com/livekit/livekit-server/pull/295/commits/1a79c18e90354151965c7029d4a4a8ee453d7347

Comment (boks1971) [reply pkg/sfu/receiver.go]: One more :-) A test was crashing because of nil pointer dereference. The channel for the `updateStats` timer gets closed after `t.receiver` is set to `nil` - https://github.com/livekit/livekit-server/pull/295/commits/3284225d283b635c5273240b5dad6d561ce7c71a

File: pkg/sfu/streamtracker.go [hunk 4]
```diff
@@ -165,14 +184,17 @@ func (s *StreamTracker) Observe(sn uint16) {
 	atomic.AddUint32(&s.countSinceLast, 1)
 }
 
-func (s *StreamTracker) detectWorker() {
+func (s *StreamTracker) detectWorker(generation uint32) {
```
Comment (davidzhao) [inline pkg/sfu/streamtracker.go:187]: clever way of dealing with this. TIL

**Review by davidzhao — APPROVED** [review]

File: pkg/rtc/mediatrack.go [hunk 5]
```diff
@@ -434,6 +446,8 @@ func (t *MediaTrack) AddReceiver(receiver *webrtc.RTPReceiver, track *webrtc.Tra
 		if t.Kind() == livekit.TrackType_AUDIO {
 			t.buffer = buff
 		}
+
+		t.startMaxQualityTimer()
```
[the remark this answers is not in the document]
Comment (boks1971) [reply pkg/rtc/mediatrack.go:450]: Will add it @davidzhao before merging.
